spec/safe_statements_spec.rb

```diff
@@ -1263,6 +1281,552 @@ def up
+                  using: :btree,
+                )
+              end
+            end
+          end
+
+          it "does not error when index exists and if_not_exists is true" do
+            create_range_partitioned_table(:foos3, migration_klass, with_partman: true)
+
+            setup_migration = Class.new(migration_klass) do
+              def up
+                safe_add_concurrent_partitioned_index :foos3, :updated_at, name_suffix: "updated_at_idx"
+              end
+            end
+
+            setup_migration.suppress_messages { setup_migration.migrate(:up) }
+
+            test_migration = Class.new(migration_klass) do
+              def up
+                safe_add_concurrent_partitioned_index :foos3, :updated_at, name_suffix: "updated_at_idx", if_not_exists: true
+              end
+            end
+
+            allow(ActiveRecord::Base.connection).to receive(:execute).and_call_original
+
+            test_migration.suppress_messages { test_migration.migrate(:up) }
+
+            aggregate_failures do
+              expect(ActiveRecord::Base.connection).to have_received(:execute).with(/CREATE INDEX IF NOT EXISTS "foos3_updated_at_idx" ON ONLY/).once
+              expect(ActiveRecord::Base.connection).to have_received(:execute).with(/CREATE INDEX CONCURRENTLY IF NOT EXISTS/).exactly(10).times
+              expect(ActiveRecord::Base.connection).to_not have_received(:execute).with(/ALTER INDEX/)
+            end
+          end
+
+          it "creates valid index when table / index name use non-standard characters" do
+            create_range_partitioned_table("foos3'", migration_klass)
+
+            # partman does not allow table names with non-standard characters
+            # so we need to create a child partition manually
+            setup_migration = Class.new(migration_klass) do
+              def up
+                unsafe_execute(<<~SQL)
+                  CREATE TABLE "foos3'_child" PARTITION OF "foos3'"
+                  FOR VALUES FROM ('2020-01-01') TO ('2020-02-01')
+                SQL
+              end
+            end
+
+            setup_migration.suppress_messages { setup_migration.migrate(:up) }
+
+            test_migration = Class.new(migration_klass) do
+              def up
+                safe_add_concurrent_partitioned_index "foos3'", :updated_at, name_suffix: "bar\""
+              end
+            end
+
+            allow(ActiveRecord::Base.connection).to receive(:execute).and_call_original
+
+            test_migration.suppress_messages { test_migration.migrate(:up) }
+
+            aggregate_failures do
+              expect(ActiveRecord::Base.connection).to have_received(:execute).with(/CREATE INDEX "foos3'_bar""" ON ONLY/).once
+              expect(ActiveRecord::Base.connection).to have_received(:execute).with(/CREATE INDEX CONCURRENTLY "foos3'_child_bar""" ON/).once
+              expect(ActiveRecord::Base.connection).to have_received(:execute).with(/ALTER INDEX/).once
```

Comment:
This should probably match `ATTACH` also.

Reply:
Done in b22c05d